[feat] 유저서비스 회원가입, 마이프로필 기능 개발

ryudain05 opened the pull request:
**관련 이슈**
🙆🏻‍♀️ [#2 유저서비스 회원가입](https://github.com/Saeongjima/danjam/issues/2)
🙆🏻‍♀️ [#3 마이페이지 RUD](https://github.com/Saeongjima/danjam/issues/3)

---

- 회원탈퇴 기능 추가 예정입니다 -> 추가했습니다.
- 예외처리가 명확하게 되지 않았습니다. 테스트를 통한 예외처리 추가 예정입니다.
- 권한 처리가 아직 추가되지 않았습니다.

ryudain05 committed:
[init] gitignore fix

ryudain05 committed:
[feat] User Entity, Role enums add

ryudain05 committed:
[feat] UserRepository add

ryudain05 committed:
[feat] JpaAuditing Cinfig add

ryudain05 committed:
[refactor] User domain Refactor
Builder와 AllArgsConstructor를 혼용하여 사용하는 경우 생성자가 2번 생성되는 문제가 야기
생성자 단에서 Builder를 사용하여 안전하게 사용하도록 수정

ryudain05 committed:
[refactor] User domain 무분별한 @Column 어노테이션 삭제

ryudain05 committed:
[feat] MyProfile Domain Add

ryudain05 committed:
[refactor] MyProfile Domain 필수 파라미터가 아닌 것들 수정
private 삭제

ryudain05 committed:
[refactor] Myprifle Directory Fix

ryudain05 committed:
[feat] School Domain Add

ryudain05 committed:
[Fix] Domain Id Jpa에서 관리함으로 생성자에서 삭제

ryudain05 committed:
[feat] JoinDto Add
회원가입 Dto 생성

ryudain05 committed:
[feat] MultiPart Util Add

ryudain05 committed:
[feat] Minio Service Add

[목적] 회원가입 처리 시 이미지 업로드를 위한 minio S3 파일 업로드, 다운로드, 삭제 기능을 위한 서비스 개발

ryudain05 committed:
[feat] MultipartUtil Add

[목적] Multipart로 파일 업로드 시 공통적으로 사용하는 Util 개발

ryudain05 committed:
[fix] Domain Fix
[목적] @GenerateValue를 지정해주지 않아 시퀀스를 생성함 -> GenerationType.IDENTITY를 지정하여 시퀀스 삭제

ryudain05 committed:
[docs] MinioService annotation Fix

ryudain05 committed:
[fix] MinioService Exception Fix
[목적] 여러 Exception 을 최상위 Exception 을 통해 코드 간결화

ryudain05 committed:
[feat] UserService Add
[목적] 유저 회원가입을 위한 서비스 생성 및 개발
학교 저장을 위한 학교 도메인 및 레포, 예외처리 생성
필요한 의존성 추가

ryudain05 committed:
[feat] 유저 회원가입 기능 개발 [fix] Minio 서버 설정으로 인한 주석 처리 및 확장자 검사 로직 수정
[목적] 유저 회원가입 기능을 위해 비밀번호 암호화 실행 -> Security 사용 및 설정
[목적] Minio Server 셋팅이 아직 안되어있음으로 주석 처리 및 확장자 선택 시 yml에서 List 값을 읽지 못하는 문제 발생 -> 따로 배열에 선언 추후 리팩토링

ryudain05 committed:
[feat] GlobalExceptionHandelr Add
[목적] ApiResponseError 형식을 통한 예외 처리를 위한 핸들러 추가
공통 예외와 @Valid만 추가,  이후에 모든 Exception 처리에 대한 핸들러 추가 예정

ryudain05 committed:
[feat] MyProfileDto Add
[목적] 마이프로필 정보 확인을 위한 Dto 생성

ryudain05 committed:
[feat] MyProfile 조회 기능 Add

ryudain05 committed:
[fix] static import fix

ryudain05 committed:
[fix] 회원가입 로직 수정

[목적] 사용자 이름을 추가하기 위한 회원가입 DTO, DOMAIN 수정
consumes mediatype을 스프링에서 지원해주는 http 값으로 수정
file 업로드 시 필수값이 아니도록 수정

ryudain05 committed:
[feat] 회원가입 중복 체크 기능 ADD

[목적] 회원가입 시 아이디, 이메일, 닉네임에 대한 중복체크. 사용할 수 있는 항목이면 true 반환, 이미 있는 항목이면 false 반환 (사용불가능)

ryudain05 committed:
[feat] OpenFeign 요청을 위한 mate dto ADD

[목적] 외부 서비스 동기 요청 시 필요한 response dto 형식 추가

ryudain05 committed:
[feat] GlobalExceptionHandler Add

[목적] 유저서비스, 마이프로필 서비스 예외처리를 위한 ExceptionHandler 설정

ryudain05 committed:
[refactor] dto 이름 DTO로 수정

ryudain05 committed:
[feat] MyProfile 정보 읽기 ADD
[목적] 마이프로필 정보 조회를 위한 API 개발
외부 서비스에서 마이프로필 정보 조회 로직 추가 예정

ryudain05 committed:
[feat] MateProfile OepngFeign 으로 가져오기
[목적] 마이프로필에서 MateProfile항목을 가져오기 위한 설정 추가

ryudain05 committed:
[refactor] MultipartUtil 책임 분리
[목적] JoinService 에 들어가 있던 파일 업로드 부분을 MultipartUtil 로 수정하여 단일책임원칙 위반 내역 수정

ryudain05 committed:
[feat] 유저 마이프로필 로그인 정보 수정 API ADD
[목적] 마이프로필 username, profileImg, password 수정을 위한 API

ryudain05 committed:
[refactor] MultipartUtil의 bucketName 메소드에 넘기도록 수정
[목적] 다른 버킷에 저장하기 위해 기존의 하드코딩 되어있던 bucketName을 파라미터로 넘기도록 수정

ryudain05 committed:
[feat] 마이프로필 학교 정보 수정 API ADD
[목적] 사용자 마이프로필의 학교 정보를 수정하기 위한 API 추가

ryudain05 committed:
[feat] 마이프로필 API 수정 및 추가
[목적] SRP 원칙 위반을 수정하기 위한 Repository 전면  수정, 마이프로필 자기소개 수정을 위한 API 개발
[달성도] 90 정확한 테스트 및 예외처리 추가 에정

ryudain05 committed:
[feat] 유저 회원탈퇴 기능 추가
[목적] soft-delete를 사용한 회원탈퇴
[달성도] 90

ryudain05 committed:
[fix] 마이프로필 수정 오류 수정
[목적] 마이프로필 로그인 정보 수정 오류 해결 및 Exception 처리 추가
[달성도] 100

kim946509 commented on src/backend/user-service/src/main/java/site/danjam/mate/user_service/domain/myProfile/controller/MyProfileController.java:
```diff
@@ -0,0 +1,68 @@
+        return ResponseEntity.ok(ApiResponseData.of(myProfileInfoService.readMyProfileInfo(username), "마이프로필 조회 성공"));
+    }
+
+    @PatchMapping(value = "/login", consumes = {MediaType.APPLICATION_JSON_VALUE,
```
consumes 속성을 사용해서 Content-type을 엄격하게 제한하신 이유가 궁금합니다! (단순 궁금)

ryudain05 replied:
아무래도 명시성때문인 거 같습니다 !

MultiPart file 데이터를 받는 API외에는 json 데이터로 받는 경우가 많기도 하여 구분 짓기 위해 작성하었습니다.

결론적으로는 불필요한 요청을 사전에 막고, Restful api 설계를 위해서 작성한 거 같습니다 ㅎㅎ

kim946509 commented on src/backend/user-service/src/main/java/site/danjam/mate/user_service/domain/myProfile/service/MyProfileInfoService.java:
```diff
@@ -0,0 +1,111 @@
+            if (userRepository.existsByUsername(dto.getUsername())) {
+                throw new DuplicateUsernameException();
+            }
+            user.updateUsername(username);
```
username이 수정되면 기존의 access&refresh 토큰 정보도 바뀌어야 할텐데 혹시 생각해보신 방법이 있으실까요?

ryudain05 replied:
이 건에 대해서는 좀 더 생각을 해봐야할 거 같습니다 ! redis를 통해 블랙리스트 기능을 만들어서 활용해도 될 거 같다는 생각을 해봤습니다... 
보안적으로는 로그아웃이 좋을 거 같지만, 사용자 경험쪽으로는 재발급이 좋을 거 같습니다..ㅜㅜ

kim946509 commented on src/backend/user-service/src/main/java/site/danjam/mate/user_service/domain/school/repository/SchoolRepository.java:
```diff
@@ -0,0 +1,16 @@
+
+@Repository
+@RequiredArgsConstructor
+public class SchoolRepository {
```
혹시 service코드도 Repository와 동일하게 클래스명을 SchoolRepository로 하신 이유가 있을까요?

ryudain05 replied:
혹시 어떤 service 코드를 말씀하시는 걸까요 ?! 제가 어떤건지 헷갈려서요 !

kim946509 commented on src/backend/user-service/src/main/java/site/danjam/mate/user_service/global/util/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,78 @@
+    public ResponseEntity<ApiResponseError> handlerInvalidInputException(InvalidInputException e) {
+        return processCustomErrorResponse(e.getErrorCode(), e.getMessage());
+    }
+
```
저희가 작성한 예외들은 모두 BaseException을 상속 받고있으니 BaseException만 예외처리 로직을 작성하는 건 어떨까요?
`@ExceptionHandler(BaseException.class)
public ResponseEntity<ApiResponseError> handleBaseException(BaseException e) {
    return processCustomErrorResponse(e.getErrorCode(), e.getMessage());
}`

ryudain05 replied:
너무 좋은 방법인 거 같습니다 !

수정하고, 반영하겠습니다 👏🏻👏🏻

ryudain05 committed:
[refactor] 프로젝트 버전 3.4.1로 변경
[목적] 프로젝트 버전을 팀원들과 맞추기 위한 설정 추가, 기존의 명시되어있던 openfeign 의 버전 문제로 인해 다운그레이드했던 버전을 dependencyManagement를 추가함으로서 수정

ryudain05 committed:
[refactor] Exception 처리 시 Code static import
[목적] 가독성을 높이기 위해 Exception 에서 사용하는 Code 클래스에 있는 공통 에러 코드에 대해서 static import 사용

High-Quality-Coffee commented on src/backend/user-service/src/main/java/site/danjam/mate/user_service/domain/myProfile/service/MyProfileInfoService.java:
```diff
@@ -98,4 +98,14 @@ public void updateGreeting(String username, GreetingDTO dto) {
         myProfileRepository.save(myProfile);
     }
+
+    @MethodDescription(description = "회원을 탈퇴합니다.")
+    public void deleteUser(String username, UpdateLoginDTO dto) {
+        User user = userRepository.findByUsername(username);
+        if (!bCryptPasswordEncoder.matches(dto.getPassword(), user.getPassword())) {
+            throw new InvalidInputException("비밀번호가 일치하지 않습니다.");
+        }
+        user.deleteUser();
+        userRepository.save(user);
+    }
```
갑자기 궁금해졌는데,
혹시 단잠은 회원이 실수로 계정을 delete 했을 경우, 복구 방법이 있을까요?

ryudain05 replied:
단잠 계정 삭제 시 soft-delete(논리 삭제)를 진행하기 때문에 계장 복구를 위하는 사용자가 있다면 복구가 가능 할 것 같습니다 !

High-Quality-Coffee reviewed (approved):
코드 구경 잘했습니닷
근데 이렇게 리뷰 남기는거 맞나ㅏ..?
